larpandoracontent/LArControlFlow/TrackDirectionTool.cc

```diff
@@ -0,0 +1,1154 @@
+
+void TrackDirectionTool::FillHitChargeVector(const Cluster *const pCluster, HitChargeVector &hitChargeVector)
+{
+    OrderedCaloHitList orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+    CaloHitList caloHitList;
+    orderedCaloHitList.FillCaloHitList(caloHitList);
+
+    const TwoDSlidingFitResult &slidingFit(this->GetCachedSlidingFit(pCluster));
+
+    for (CaloHitList::const_iterator hitIter = caloHitList.begin(), hitIterEnd = caloHitList.end(); hitIter != hitIterEnd; ++hitIter)
+    {
+        const CaloHit *const pCaloHit(*hitIter);
+        const CartesianVector caloHitPosition(pCaloHit->GetPositionVector());
+        float hitWidth(pCaloHit->GetCellSize1());
+
+        float caloHitEnergy(pCaloHit->GetInputEnergy());
+        caloHitEnergy *= 273.5; //ADC to electron
+        caloHitEnergy *= 23.6/1000000; //ionisation energy per electron in MeV
+        caloHitEnergy /= 0.62;
+
+        float rL(0.f), rT(0.f);
+        slidingFit.GetLocalPosition(caloHitPosition, rL, rT);
+        if (rL == 0.)
+            continue;
+
+        float calibratedUncertainty(std::sqrt((0.00419133 * (caloHitEnergy/hitWidth) * (caloHitEnergy/hitWidth)) + (0.00967141 * (caloHitEnergy/hitWidth)))); //70%
+        HitCharge hitCharge(pCaloHit, rL, hitWidth, caloHitEnergy, calibratedUncertainty);
+        hitChargeVector.push_back(hitCharge);
+    }
+
+    std::sort(hitChargeVector.begin(), hitChargeVector.end(), SortHitChargeVectorByRL);
+
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::TrackInnerFilter(HitChargeVector &hitChargeVector, HitChargeVector &filteredHitChargeVector)
+{
+    //Fill endpoint protected area into filtered vector and put all other hits in a separate vector
+    float endpointProtectionRange(0.05);
+    filteredHitChargeVector.insert(filteredHitChargeVector.begin(), hitChargeVector.begin(),  hitChargeVector.begin() + endpointProtectionRange * hitChargeVector.size());
+    filteredHitChargeVector.insert(filteredHitChargeVector.begin(), hitChargeVector.begin() + (1.0 - endpointProtectionRange) * hitChargeVector.size(), hitChargeVector.end());
+
+    HitChargeVector innerHitChargeVector(hitChargeVector.begin() + endpointProtectionRange * hitChargeVector.size(), hitChargeVector.begin() + (1.0 - endpointProtectionRange) * hitChargeVector.size());
+
+    int nNeighboursToConsider(5);
+    this->SetNearestNeighbourValues(innerHitChargeVector, nNeighboursToConsider);
+
+     std::sort(innerHitChargeVector.begin(), innerHitChargeVector.end(), SortByDistanceToNN);
+     filteredHitChargeVector.insert(filteredHitChargeVector.begin(), innerHitChargeVector.begin(), innerHitChargeVector.begin() + 0.72 * innerHitChargeVector.size()); //lots of testing has been done to optimise percentage
```

Comment:
But it is an optimised number for it to work best? This is Joris' thinking, but it will be due to the massively varying list sizes. The tail effects will thus kick in at different places.

Reply:
Having this as the default value for MicroBooNE based on the optimisation is fine, but the optimal value may be different for different use cases.

Reply:
Yes, done